samples/navigate-route-with-rerouting/src/main/java/com/esri/arcgismaps/sample/navigateroutewithrerouting/components/NavigateRouteWithReroutingViewModel.kt

```diff
@@ -0,0 +1,400 @@
+
+    // Graphics overlay to display the route ahead and traveled graphics
+    val graphicsOverlay = GraphicsOverlay()
+
+    // Keep track of the the location display job when navigation is enabled
+    private var locationDisplayJob: Job? = null
+
+    // Default location display object, which is updated by rememberLocationDisplay
+    private var locationDisplay: LocationDisplay = LocationDisplay()
+
+    // Starting point: San Diego Convention Center
+    private val startingStop = Stop(Point(-117.160386, 32.706608, SpatialReference.wgs84()))
+        .apply { name = "San Diego Convention Center" }
+
+    // Destination point: Fleet Science Center
+    private val destinationStop = Stop(Point(-117.146679, 32.730351, SpatialReference.wgs84()))
+        .apply { name = "RH Fleet Aerospace Museum" }
+
+    // Generate a route with directions and stops for navigation
+    val routeTask = RouteTask(
+        pathToDatabase = "$provisionPath/sandiego.geodatabase",
+        networkName = "Streets_ND"
+    )
+
+    // The route parameters needed to calculate a route from a start and stop point
+    private var routeParameters = RouteParameters()
+
+    // The resulted route from the route task using the route parameters
+    private var routeResult: RouteResult? = null
+
+    // Instance of the route ahead polyline
+    private var routeAheadGraphic: Graphic = Graphic(
+        symbol = SimpleLineSymbol(
+            SimpleLineSymbolStyle.Dash,
+            Color(getColorArgb(com.esri.arcgismaps.sample.sampleslib.R.color.colorPrimary)),
+            3f
+        )
+    )
+
+    // Graphic to represent the route that's been traveled (initially empty)
+    private var routeTraveledGraphic: Graphic = Graphic(
+        symbol = SimpleLineSymbol(
+            SimpleLineSymbolStyle.Solid,
+            Color.black,
+            3f
+        )
+    )
+
+    var distanceRemainingText by mutableStateOf("")
+        private set
+
+    var timeRemainingText by mutableStateOf("")
+        private set
+
+    var nextDirectionText by mutableStateOf("")
+        private set
+
+    var isNavigateButtonEnabled by mutableStateOf(true)
+        private set
+
+    var isRecenterButtonEnabled by mutableStateOf(false)
+        private set
+
+    // Boolean to check if Android text-speech is initialized
+    private var isTextToSpeechInitialized = AtomicBoolean(false)
+
+    // Instance of Android text-speech
+    private var textToSpeech: TextToSpeech? = null
+
+    // JSON of polylines of the path for the simulated data source
+    private val polylineJSON: String by lazy {
+        application.getString(R.string.simulation_path_json)
+    }
+
+    // Polyline representing simulation path
+    private val simulationPolyline = Geometry.fromJsonOrNull(polylineJSON) as Polyline
+
+    // Create a ViewModel to handle dialog interactions
+    val messageDialogVM: MessageDialogViewModel = MessageDialogViewModel()
+
+    fun initialize(locationDisplay: LocationDisplay) {
```

Comment:
I saw this pattern in the other sample too...what is the advantage of this over an init block?

Reply:
If I'm correct, the main difference is `init` is triggered during the ViewModel constructor invocation, and using this pattern guarantees this function is called after the ViewModel constructor invocation. The advantage here is we are not spawning coroutines from the `init` and ensures that the the ViewModel contains an asynchronous initialization. (until ViewModel can trigger our `initialize()`)